### ISSUE TITLE
Cygwin shell interfering on Windows

### ISSUE THREAD

**Issue by dutchgecko** [airblade/vim-gitgutter#285]
Hi,

I've been having this issue for a while but have only now had time to look into it. When I open a file in Vim for Windows, vim-gitgutter produces an E484 error. Having looked into it, it seems a command prompt is spawned in the background, which is filled with error messages:

![](http://i.imgur.com/2MJSgiP.png)

The prompt closes as soon as the last errors are produced, but by running vim from the command line I've managed to capture those too:

![](http://i.imgur.com/sa41ZfJ.png)

To be honest I don't know at all what's going on, but if I had to guess it's that the windows version of vim is picking up the cygwin version of sh that's installed on my computer.

Some useful things to note:
vim in cygwin works absolutely fine
The grep test can be run from a windows command prompt and will succeed (forward slashes and all)
While git is also installed in cygwin, the version available in PATH is the windows version
Whenever the error occurs vim will freeze for a few moments. The background command prompt has small pauses between its output. A file called sh.exe.stackdump is produced in vim's working directory (if writeable).
Vim, vim-gitgutter and cygwin are all updated to the latest version.

**Comment by airblade** [comment]
The E484 simply means Vim was unable to run an external command successfully.

I'm not sure whether the running of external commands is failing because the commands have the wrong syntax or because there's some unrelated problem with running external commands in general.

I googled `Win32 error 487` from your screenshot and found a [StackOverflow](http://stackoverflow.com/q/18502999/151007) question.  Do any of the suggestions there help?

You could also try installing vim-misc and vim-shell (see the [README](https://github.com/airblade/vim-gitgutter#installation)) instead.

In summary, I have no idea what's wrong...

**Comment by dutchgecko** [comment]
OK, I've performed the following changes:

MSysGit has been updated to version 2 (as per one piece of advice in the SO topic)
Cygwin binaries have been removed from PATH
MSysGit binaries have been added to PATH (basically, sh now points to the MSysGit version instead of the Cygwin version)
I've installed vim-misc and vim-shell.

I'm still getting errors, but the message has changed.

This is when opening a file:
![](http://i.imgur.com/j3EzIBW.png)

And this is during editing (whenever cursor has stopped)
![](http://i.imgur.com/P0an3uu.png)

While MSysGit does provide a shell, I'm wondering if it's providing enough other utilities. Grep for example doesn't seem to be in the bin directory. Are there any other unix commands which your plugin depends on? Other than the git ones obviously :)

**Comment by airblade** [comment]
The plugin only uses `git`.  It'll use `grep` if it's available but work fine if it's not.

You could add `echo a:cmd` after [this line](https://github.com/airblade/vim-gitgutter/blob/1aed46c072fd9bd00650474e89f0e4d3a483dad8/autoload/gitgutter/utility.vim#L108) to see what command vim-gitgutter is trying to run, then try to run that command yourself on the command line.  You'll get a more helpful error message then.

**Comment by lionelmessi** [comment]
I am also getting the same error on Windows.

**Comment by lionelmessi** [comment]
Updating to the latest airblade fixes this issue, thanks!

**Comment by dutchgecko** [comment]
Hmm, there's a few things going on. For example, this was when opening $MYVIMRC (C:\Users\gecko\_vimrc):

```
gecko@shendelzare MINGW64 ~$ cd "C:\Users\gecko" && (git ls-files --error-unmatch _vimrc && (git diff --no
-ext-diff --no-color -U0  -- _vimrc))
fatal: Not a git repository (or any of the parent directories): .git
```

I think in this case it's getting stuck on the .git global configuration directory. When opening vim, cd-ing to C:\Users\gecko\.vim (which is a git repo) and then opening vimrc in that directory, there is no problem.

However, I started getting issues again soon after. In vim, `:e bundle/vim-gitgutter/autoload/gitgutter/utility.vim` (vim-gitgutter is checked out as a submodule) causes the following command to run and fail: `cd "bundle\vim-gitgutter\autoload\gitgutter" && (git ls-files --error-unmatch utility.vim && (git diff --no-ext-diff --no-color -U0  -- utility.vim))` When running from a shell in the command prompt, that command succeeds.

Instead, I do this in vim: 
```
:cd bundle\vim-gitgutter\autoload\gitgutter\
:e utility.vim
```

No error. Something about being in the right directory helps the command to succeed.

Similarly:
```
:cd $VIMRUNTIME
:e C:\Users\gecko\.vim\vimrc
```

causes an error, but with a different command: `cd "\Users\gecko\.vim" && (git show :vimrc > C:/Users/gecko/AppData/Local/Temp/VIGF521.tmp && git diff --no-ext-diff --no-color -U0  -- C:/Users/gecko/AppData/Local/Temp/VIGF521.tmp C:/Users/gecko/AppData/Local/Temp/VIHF522.tmp || exit 0)`

Again, running from a shell allows the command to succeed.

**Comment by airblade** [comment]
What does `:echo &shell` produce?  Is it `fish`?

vim-gitgutter is supposed to wrap each command in parentheses ([opening](https://github.com/airblade/vim-gitgutter/blob/1aed46c072fd9bd00650474e89f0e4d3a483dad8/autoload/gitgutter/diff.vim#L17), [closing](https://github.com/airblade/vim-gitgutter/blob/1aed46c072fd9bd00650474e89f0e4d3a483dad8/autoload/gitgutter/diff.vim#L73)) so Windows `cd`s properly into the directory of the file in question – unless you're using fish.

In your commands the wrapping parentheses are missing; and your results suggest the `cd` isn't working.

**Comment by dutchgecko** [comment]
`:echo &shell` produces sh (which is what I expect, as I have `set shell=sh` in my vimrc).
`!echo $SHELL` opens a command prompt with result /usr/bin/bash.

It's worth noting that I do have fish installed in cygwin, but have had issues with other vim plugins in the past and so have done my best to make sure vim runs its commands under a normal shell.

**Comment by airblade** [comment]
Now I look at the code again I remember that the wrapping with parentheses doesn't encompass the initial `cd`.

vim-gitgutter always `cd`s into a file's directory before running the git commands.  You have already verified vim-gitgutter works if you manually `cd` to the directory first, and also that the generated commands work on the command line.  So it seems vim-gitgutter's `cd` isn't working properly.

One thing we haven't tried is running the generated commands in vim.  You said that in vim `:e bundle/vim-gitgutter/autoload/gitgutter/utility.vim` fails.  What happens if you do this? :-

```viml
:echo system('cd "bundle\vim-gitgutter\autoload\gitgutter" && (git ls-files --error-unmatch utility.vim && (git diff --no-ext-diff --no-color -U0 -- utility.vim))')
```

**Comment by ascarpino** [comment]
@airblade by manually running that command I get the same message "Error detected... gitgutter#utility#system:" and "Can't open C:/....tmp/xyz.tmp".

I'm running cygwin with gvim-7-4-827.

**Comment by alexhardwicke** [comment]
I have the same problem - only with gvim, and as far as I can see everyone else here is also having the problem with gvim.

    :echo &shell gives /usr/bin/bash

If I run the echo system command two comments up in gvim, I get the same error that's constantly produced:

    Error detected while processing function gitgutter#process_buffer[8]..gitgutter#diff#run_diff[13]..gitgutter#utility#file_relative_to_repo_root[3]..gitgutter#utility#system:
    line    10:
    E484: Can't open file C:/Users/Alex/AppData/Local/Temp/VIoCA10.tmp


I have no problems at all in terminal vim (I run it via mintty). If I run the echo system command, I get:

    /usr/bin/bash: line 0: cd: bundle\vim-gitgutter\autoload\gitgutter: No such file or directory

**Comment by airblade** [comment]
It seems relevant that this appears to affect only Windows gvim.  And it also seems to be related to vim's current working directory.

Unfortunately I'm at a bit of a loss: I don't use Windows so I can't investigate myself nor make any sensible suggestions.

**Comment by alexhardwicke** [comment]
I've done some research and at least on my end, I've fixed it.

I run regular vim on windows in mintty (Cygwin), and have a few problems there and there (like https://github.com/git-for-windows/git/issues/498). I solved the problems by setting my shell to /bin/bash, which fixes the command line vim problems on Windows and has no effect on my Linux machines (I use the same dotfiles everywhere).

This seems to be the culprit. It means gvim on Windows is trying to run /bin/bash, which doesn't exist because gvim on windows doesn't have the entire Cygwin environment and /bin/bash is meaningless. I've wrapped the set to make sure that gui_running is false.

To make it clear, what I've done is change

    set shell=/bin/bash

to

    if has("gui_running") == 0
        set shell=/bin/bash
    endif

Now that I've changed my vimrc to that, the shell in gvim on windows is correctly cmd.exe, and it seems to be able to make the system calls to the temp folder/files with no problems.

So my suggestion to other people reading this is to make sure the shell is properly set in gvim. Run

    :verbose set shell? shellcmdflag? shellxquote?

to see the output. You want to have something similar to the following:

    shell=C:\WINDOWS\system32\cmd.exe
    shellcmdflag=/c
    shellxquote=(

**Comment by dutchgecko** [comment]
I've followed @alexhardwicke's instructions and ensured that shell is not set in the GUI. This fixes the situation for me too.

> This seems to be the culprit. It means gvim on Windows is trying to run /bin/bash, which doesn't exist because gvim on windows doesn't have the entire Cygwin environment and /bin/bash is meaningless. 

This isn't quite true in my case: shell was set to `sh` which was on the path (git's bash shell). I suspect that vim was constructing a windows-style command and sending it to sh, which of course causes problems.

Thanks everybody for your help. @airblade, do you want to investigate this issue further to see if it can be patched in gitgutter, or do you consider this a required workaround (seems reasonable to me)?

**Comment by alexhardwicke** [comment]
@dutchgecko Could be. The shellcmdflag/shellxquote flags are supposed to help vim form the commands properly, but if they weren't correct for sh (or if vim's OS detection just doesn't work for a posix shell on Windows) I'd not be surprised.

**Comment by alexhardwicke** [comment]
As for fixing the issue in gitgutter, once I suspected it was shell I noticed that my shell was being modified by syntastic. That seems to have had the same problem, and have fixed it by having a wrapper around system that fiddles with the shell, saving the original value locally and setting it back afterwards:

https://github.com/scrooloose/syntastic/blob/d1a179d750bd1d136d7f38e69f2c5b8439886de7/autoload/syntastic/util.vim#L31

So it is possible to patch it, but requiring a workaround seems reasonable enough to me, personally.

**Comment by airblade** [comment]
@alexhardwicke Sounds like you've found the problem.

Vim tries to detect the user's shell and adjusts various options accordingly so external commands work.  I don't want to try to second-guess Vim's detection.

So I'll leave the plugin as-is and if similar problems come up in future we'll know where to look.
